### PR TITLE
Enable load-store reduction again

### LINKED ISSUE #302: enable_load_reductions(graph) in -Jred throws a type_error
**Recreate**
Modify jlm/llvm/opt/reduction.cpp (line 134)

static void
reduce(RvsdgModule & rm, util::StatisticsCollector & statisticsCollector)
{

and disable all optimizations except the load_reduction, i.e.,
  // enable_mux_reductions(graph);
  // enable_store_reductions(graph);
  enable_load_reductions(graph);
  // enable_gamma_reductions(graph);
  // enable_unary_reductions(graph);
  // enable_binary_reductions(graph);

Compile the file separately to avoid having to comment out unused functions:
/usr/lib/llvm-16/bin/clang++ -c --std=c++17 -gdwarf-4 -g -DJLM_DEBUG -DJLM_ENABLE_ASSERTS -I. -I/usr/lib/llvm-16/include -I. -I/usr/lib/llvm-16/include -o build/jlm/llvm/opt/reduction.la jlm/llvm/opt/reduction.cpp

Then compile jlm with debug: 
make jlm-debug -j `nproc`

Then unzip the attached ZIP and type 'make'

**Output**
terminate called after throwing an instance of 'jlm::util::type_error'
  what():  Type error - expected : float, received : bit32
Aborted (core dumped)

[classify-f32.zip](https://github.com/phate/jlm/files/13560411/classify-f32.zip)

### PR DESCRIPTION
**Pull request by phate**
1. Fixes the load-store reduction. The problem was a missing check for the stored and loaded value type being the same. If they are not the same, then we might have to insert a truncation or conversion operation instead of just bypassing the load and store. I marked the place in the code with a FIXME such that we can improve upon it in the future.
2. Adds a unit test that exposes the problem.

Closes #302

### CODE REVIEW

File: jlm/llvm/ir/operators/load.cpp [hunk 1]
```diff
@@ -197,30 +197,67 @@ is_multiple_origin_reducible(const std::vector<rvsdg::output *> & operands)
...
+  if (loadAddress != storeAddress)
+  {
+    return false;
+  }
+
+  // Check that the loaded and stored value type are the same
+  //
+  // FIXME: This is too restrictive and can be improved upon by inserting truncation or narrowing
+  // operations instead. For example, a store of a 32 bit integer followed by a load of a 8 bit
+  // integer can be converted to a trunc operation. Similarly, a store of a floating point
```
Comment (sjalander) [inline jlm/llvm/ir/operators/load.cpp:251]: Storing an FP value and then reading it back as an INT shouldn't perform any conversion, and shouldn't mean the same thing as using a machine instruction that convert FP to INT.

**Review by sjalander — DISMISSED** [review]

File: jlm/llvm/ir/operators/load.cpp (hunk 1 above)
Comment (phate) [reply jlm/llvm/ir/operators/load.cpp]: You are absolutely right. I removed the last sentence from the FIXME.

**Comment by phate** [comment]
@sjalander or @haved : I need a new approval, please.

File: jlm/llvm/ir/operators/load.cpp [hunk 2]
```diff
@@ -197,30 +197,66 @@ is_multiple_origin_reducible(const std::vector<rvsdg::output *> & operands)
   return states.size() != operands.size() - 1;
 }
 
+// s2 = store_op a v1 s1
+// v2 s3 = load_op a s2
+// ... = any_op v2
+// =>
```
Comment (haved) [inline jlm/llvm/ir/operators/load.cpp:203]: This comment kind of makes it appear as if the store disappears, I assume it doesn't?

**Review by haved — APPROVED** [review]
LGTM, just one small comment

File: jlm/llvm/ir/operators/load.cpp (hunk 2 above)
Comment (phate) [reply jlm/llvm/ir/operators/load.cpp:203]: No, the store does not disappear. The only node that is removed is the load. I clarified in the function documentation by adding the store.

**Comment by phate** [comment]
@sjalander and @haved : Clarified the documentation again after Hårvard's comment. I would need another approval.

**Review by haved — APPROVED** [review]
LGTM

**Pull request merged**